### DISCUSSION TITLE
Get CID from an array of files without uploading them

### DISCUSSION THREAD

**Discussion by jaymgonzalez** [web3-storage/web3.storage#2157] [Q&A]
Hi team, 

I'm building an app with NextJs and I need to get the CID of some files before uploading them. Is there a way of achieving this using your library or creating the hash the same you do in the library? I tried looking at your code in `https://github.com/web3-storage/web3.storage/blob/main/packages/client/src/lib.js` to find the way you get the CID but not sure I understand it correctly. 

In my app I'm using the JS library from the browser. 

I tried returning from the function bellow but it didn't work. (or it worked as expected, it returned the CID but it also uploaded the files)
```
async function getCid(files) {
    const onRootCidReady = (cid) => {
      console.log('uploading files with cid:', cid)
      return cid
    }

    const client = makeStorageClient()

    client.put(files, { onRootCidReady })
  }
  ```
Please let me know if there is a way of achieving this.

Any help is appreciated. Thanks ☺️

**Comment by toshitehdev** [comment]
I am interested to know how to do this as well. Currently I am using multiformats library, but the CID returned is different with different file size

**Reply by jaymgonzalez** [reply]
hey @toshitehdev, 

i was having the same problem but found a solution, maybe it helps you too, 

```

import { pack } from 'ipfs-car/pack'

function toImportCandidate(file) {
  let stream
  return {
    path: file.name,
    get content() {
      stream = stream || file.stream()
      return stream
    },
  }
}

export async function getCid(files) {
  try {
    const { root } = await pack({
      input: Array.from(files).map(toImportCandidate),
      wrapWithDirectory: true,
      maxChunkSize: 1048576,
      maxChildrenPerNode: 1024,
    })
    return root.toString()
  } catch (err) {
    console.log(err)
  }
}

``` 

passing the files to getCid function returns the correct CID for me. [here](https://github.com/jaymgonzalez/generatenfts/tree/main/front-end) is the link to the repo in case you want to investigate further.

hope it helps

**Reply by toshitehdev** [reply]
i just noticed this is web3.storage. I am using nft.storage lol, already found the solution tho
